keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+
+### Goals
+
+* Provide structured logging interface in K8s codebase
+* Standardize commonly used logging keys (e.g. referencing k8s object, http request fields)
+* Propose sound strategy for migration ensuring its completeness
+* Make implementation of context based solutions like logs correlation or tracing within Kubernetes easier.
+
+### Non-Goals
+
+* We are **not** changing logging pipeline in K8s
+* We are **not** standardizing some specific log messages
+* We are **not** tackling logging keys lifecycle (should be addressed by separate KEP when new API is close to GA)
+
+## Proposal
+
+This KEP proposes introducing a structured logging into Kubernetes control-plane components. As part of this proposal we will introduce a new logging API that will replace klog. We will introduce a Kubernetes logging schema which will provide a way to define a common set of logging keys and ensure their consistent usage. To provide immediate benefit for adapting a new API we will propose a new logging format that will simplify ingestion of structured data into third party logging solutions. As part of changes in kubernetes logging we propose to tightly couple usage of logging and golang context.
+
+### Logging Interface
+
+When choosing new structured logging API following criteria has been used:
+* Selected API should provide abstract structured logging interface so that it would be easy to inject different implementations for the same API.
+* Selected API should be minimalistic so that it should be easy to reimplement, wrap or to mock it.
+* Applicability of the selected API should not be limited to the Kubernetes control plane but it should be possible to use the same API in other components running on Kubernetes clusters which are also implemented in go.
+* Selected API should support structured logging by providing a way to define a set of key-value pairs for each log entry.
+* There should at least one production ready implementation of the selected API which could produce JSON formatted logs.
+* It should be possible to produce klog compatible logs from the selected API which could be used during the migration to a new API.
+
+Based on this criteria https://github.com/go-logr/logr log interface has been selected as a logging API for Kubernetes.
```

Comment:
A table listing alternatives would help. I'm personally a fan of log libraries with pluggable formatters, filters, and handlers to give flexibility in how to serialize, what is serialized, and full override.  In that world, json is a formatter, turning on for a conditional is a filter, and log exfiltration is a handler.

Reply:
I tend to agree with @deads2k and @brancz that before jumping to a conclusion, it would help the reader know what other options were reviewed and discarded.

Reply:
I agree that all of those are important generic logging library features but most of those are used only during the library initialization/configuration and they are mostly irrelevant from the perspective of the developer who simply wants to log something.
Therefore I believe that the usage of simple logging API like logr throughout the Kubernetes components code which is underneath backed up by logging library like zap with all the features which you have mentioned initialized during the component startup aligns nicely with the need of simplicity and replicability from one side (logr) and the flexibility and configurability from the other (zap).

Reply:
We have decided to tackle improving klog log format first.

Reply:
Picking a logging library is topic that will require dedicated discussion. We removed it from scope of this KEP.